### PR TITLE
[BugFix][Spec Decode] Preserve RNG state for zero-draft requests

### LINKED ISSUE #13608: [Bug]: Zero-draft requests incorrectly advance per-request RNG in rejection sampling
### Your current environment

<details>
<summary>Relevant environment and revisions</summary>

```text
vLLM: 0.25.1 (752a3a504485790a2e8491cacbb35c137339ad34)
vLLM Ascend affected release revision: 50e0ce60849a7b6fe6f94e1de085b6fc06e7006d
vLLM Ascend main: bdab23475a7dc72333d0c51a2bbc8ebbecb97b69
PyTorch: 2.10.0
torch-npu: 2.10.0.post2
NPU: Ascend 910B3
```

The bug is also visible by code inspection and a generator-state unit test; it is not hardware-specific.

</details>

### 🐛 Describe the bug

`sample_recovered_tokens()` advances a request's seeded generator even when that request has no draft tokens.

Current code:

```python
num_draft_tensor = torch.tensor(num_draft_tokens, pin_memory=True).to(
    device, non_blocking=True
)
has_draft_mask = num_draft_tensor > 0

for i, generator in sampling_metadata.generators.items():
    temp_q = torch.empty_like(q[i])
    temp_q.exponential_(generator=generator)
    q[i] = torch.where(has_draft_mask[i], temp_q, q[i])
```

When `num_draft_tokens[i] == 0`, `torch.where` discards `temp_q`, but `temp_q.exponential_(generator=generator)` has already advanced the per-request RNG state by a full vocabulary-sized draw.

This is a regression introduced by #4587. Before that vectorization, the implementation explicitly preserved the generator state for zero-draft requests:

```python
for i, generator in sampling_metadata.generators.items():
    # Do not generate random numbers for requests with no draft tokens.
    # This can be important for reproducibility.
    if num_draft_tokens[i] > 0:
        q[i].exponential_(generator=generator)
```

Upstream vLLM still has this guard.

A minimal state-level reproducer for the current behavior is:

```python
import torch

device = torch.device("npu")
generator = torch.Generator(device=device).manual_seed(1234)
state_before = generator.get_state().clone()

q = torch.empty(128, dtype=torch.float32, device=device)
q.exponential_()

has_draft = torch.tensor(False, device=device)
temp_q = torch.empty_like(q)
temp_q.exponential_(generator=generator)
q = torch.where(has_draft, temp_q, q)

assert torch.equal(state_before, generator.get_state())
# Fails: the discarded exponential samples advanced the generator.
```

Expected behavior:

- A request with zero draft tokens must leave its per-request generator state unchanged.
- A request with one or more draft tokens must advance its generator state.

Actual behavior:

- Both zero-draft and active-draft requests advance their per-request generators.

### Impact

Zero-draft rounds can depend on scheduling and batch composition. Consequently, later rejection sampling for the same seeded request starts from a different RNG state depending on concurrency and scheduling, changing the generated token sequence and breaking fixed-seed reproducibility.

This does not by itself prove a biased marginal sampling distribution. It is a seed/reproducibility correctness issue: speculative decoding should not consume request RNG for a sampling operation that did not occur.

We observed scheduling-dependent output changes with DSpark on a 16-NPU 1P1D deployment. This may contribute to the behavior in #13441, whose affected revision and default DSpark probability-space path contain this bug, but direct causality for that issue is not yet established because its raw token IDs and zero-draft trace are unavailable.

### Suggested fix

Restore the original semantic guard:

```python
for i, generator in sampling_metadata.generators.items():
    if num_draft_tokens[i] > 0:
        q[i].exponential_(generator=generator)
```

Please add a regression test that verifies:

1. `num_draft_tokens[i] == 0` leaves the generator state unchanged.
2. `num_draft_tokens[i] > 0` advances the generator state.

Related: #4587, #13441

### PR DESCRIPTION
**Pull request by yaleyoou**
### What this PR does / why we need it?

  `sample_recovered_tokens()` consumed per-request RNG for zero-draft
  requests and then discarded the generated exponential samples with
  `torch.where`. This made later rejection sampling depend on scheduling
  and batch composition.

  Use the existing Python `num_draft_tokens` data to skip generator-aware
  draws for zero-draft rows. Active-draft and unseeded sampling behavior
  remain unchanged. Add NPU regression coverage for zero, active, mixed,
  independent, partial-generator, and skipped-round cases.

  Fixes #13608

  ### Does this PR introduce _any_ user-facing change?

  No API change. Fixed-seed speculative decoding now preserves request RNG
  state during zero-draft rounds.

  ### How was this patch tested?

  - 7 new focused NPU regression tests passed
  - 31 existing NPU rejection-kernel tests passed
  - 19 rejection-sampler unit tests passed
  - 85 spec-decode unit tests passed
  - No scalar synchronization, new H2D copy, or persistent HBM growth was observed
  - `git diff --check` and relevant pinned formatting/lint hooks passed
  - Full `format.sh ci` did not complete because actionlint's Go dependency
    download timed out at the proxy
  - Ascend 910B3, CANN 9.0.0, torch 2.10.0,
    torch-npu 2.10.0.post2, triton-ascend 3.2.1
    https://github.com/vllm-project/vllm/commit/0351e9aa1fdf1a51329d1906881528dfe61fc88e

- vLLM version: v0.27.1
- vLLM main: https://github.com/vllm-project/vllm/commit/cdc4824a21eaa986d4d1fee90a7e6465c9f706e6

### CODE REVIEW

**Comment by yaleyoou** [comment]
@realliujiaxu @wangxiyuan @Yikun  Could one of you please add the ready-precise label? Test selection has completed and cigate is waiting for it.
